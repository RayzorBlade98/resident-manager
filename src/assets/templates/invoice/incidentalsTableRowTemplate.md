<tr>
    <td>{{NAME}}</td>
    <td>{{TOTAL_COST}}</td>
    <td>{{DEDUCTION_TYPE}}</td>
    <td>{{RESIDENT_COST}}</td>
</tr>
