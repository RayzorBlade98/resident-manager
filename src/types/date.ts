/**
 * Enum containing all months as `string`
 */
export enum Month {
  January = 'Januar',
  Febuary = 'Februar',
  March = 'MÃ¤rz',
  April = 'April',
  May = 'Mai',
  June = 'Juni',
  July = 'Juli',
  August = 'August',
  September = 'September',
  October = 'Oktober',
  November = 'November',
  December = 'Dezember',
}

/**
 * Type representing a year
 */
export type Year = number;

/**
 * Information about a specific month and year
 */
export interface MonthYear {
  month: Month;
  year: Year;
}

/**
 * Type of the string representation of a `MonthYear`
 *
 * Format: "[month] [year]"
 */
export type MonthYearString = `${Month} ${Year}`;

/**
 * Type of the string representation of a date including day, month and year
 *
 * Format: `dd.mm.yyyy`
 */
export type DateString = `${string}.${string}.${string}`;

/**
 * Type of the date string representation returned by the date input field
 *
 * Format: `yyyy-mm-dd`
 */
export type DateInputString = `${string}-${string}-${string}`;

/**
 * Class that provides utility functions for `MonthYear` objects
 */
export abstract class MonthYearUtils {
  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param monthYear `MonthYear` object to which the months should be added
   * @param addedMonths Number of months that should be added (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static addMonths(monthYear: MonthYear, addedMonths = 1): MonthYear {
    if (addedMonths < 0) {
      // Add negative amount of months == subtract months
      return MonthYearUtils.subtractMonths(monthYear, addedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.December) {
      // Year change
      newMonthYear.month = Month.January;
      newMonthYear.year += 1;
    } else {
      // Add one month to the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex + 1];
    }

    // Call `addMonths` recursive until all months got addded
    if (addedMonths !== 1) {
      return MonthYearUtils.addMonths(newMonthYear, addedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Checks if two `MonthYear` objects are equal
   * @param a first `MonthYear` object to compare
   * @param b second `MonthYear` object to compare
   * @returns `true` if both the month and year are the same, else `false`
   */
  public static areEqual(a: MonthYear, b: MonthYear): boolean {
    return a.month === b.month && a.year === b.year;
  }

  /**
   * Compares two `MonthYear` objects
   * @param a First `MonthYear` object to compare
   * @param b Second `MonthYear` object to compare
   * @returns
   *
   * - 0 if both object have the same month and year
   * - <0 if the first object contains a date that's earlier than the second object
   * - \>0 if the first object contains a date that's later than the second object
   */
  public static compare(a: MonthYear, b: MonthYear): number {
    if (a.year !== b.year) return a.year - b.year;

    const months = Object.values(Month);
    const aMonthIndex = months.findIndex((m: Month) => m === a.month);
    const bMonthIndex = months.findIndex((m: Month) => m === b.month);
    return aMonthIndex - bMonthIndex;
  }

  /**
   * Returns a `MonthYear` object that contains information of the current month and year
   * @returns `MonthYear` object for the current date
   */
  public static getCurrentMonthYear(): MonthYear {
    const date: Date = new Date();
    const monthIndex: number = date.getUTCMonth();
    const month: Month = Object.values(Month)[monthIndex] as Month;
    const year: Year = date.getUTCFullYear();

    return {
      month,
      year,
    };
  }

  /**
   * Parses a string representation of a `MonthYear` object to its respective object
   * @param monthYear string representation of a `MonthYear` object that should be parsed
   * @returns `MonthYear` object that contains the information of the specified `string`
   */
  public static parseString(monthYear: MonthYearString): MonthYear {
    const splits = monthYear
      .split(/(\s+)/)
      .filter((s: string) => s.trim().length > 0);
    return {
      month: splits[0] as Month,
      year: Number(splits[1]),
    };
  }

  /**
   * Subtracts the specified number of months from the given `MonthYear` object
   * @param monthYear `MonthYear` object from which the months should be subtraced
   * @param subtractedMonths Number of months that should be subtracted (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static subtractMonths(
    monthYear: MonthYear,
    subtractedMonths = 1,
  ): MonthYear {
    if (subtractedMonths < 0) {
      // Subtract negative amount of months == add months
      return MonthYearUtils.addMonths(monthYear, subtractedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.January) {
      // Year change
      newMonthYear.month = Month.December;
      newMonthYear.year -= 1;
    } else {
      // Subtract one month from the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex - 1];
    }

    // Call `subtractMonths` recursive until all months got subtracted
    if (subtractedMonths !== 1) {
      return MonthYearUtils.subtractMonths(newMonthYear, subtractedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Converts a `MonthYear` to its string representation
   * @param monthYear `MonthYear` that should be converted
   * @returns string representation of the specified `MonthYear` object with the format "[month] [year]"
   */
  public static toString(monthYear: MonthYear): MonthYearString {
    return `${monthYear.month} ${monthYear.year}`;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    const timespan = [{ ...start }];

    // Whether to add a month each loop or subtract one
    const direction = Math.sign(MonthYearUtils.compare(end, start));

    if (MonthYearUtils.areEqual(start, end)) return timespan;

    for (
      let next = MonthYearUtils.addMonths(start, direction);
      !MonthYearUtils.areEqual(next, end);
      next = MonthYearUtils.addMonths(next, direction)
    ) {
      timespan.push(next);
    }
    timespan.push({ ...end });
    return timespan;
  }
}

/**
 * Class that provides utility functions for dates containing day, month and year.
 */
export abstract class DateUtils {
  /**
   * Converts a date string from `yyyy-mm-dd` to `dd.mm.yyyy` format
   * @param date Date in `yyyy-mm-dd` format
   * @returns Converted date in `dd.mm.yyyy` format
   */
  public static convertDateInputString(date: DateInputString): DateString {
    const splits = date.split('-');
    return `${splits[2]}.${splits[1]}.${splits[0]}`;
  }

  /**
   * Converts a date string from `dd.mm.yyyy` to `yyyy-mm-dd` format
   * @param date Date in `dd.mm.yyyy` format
   * @returns Converted date in `yyyy-mm-dd` format
   */
  public static convertToDateInputString(date: DateString): DateInputString {
    const splits = date.split('.');
    return `${splits[2]}-${splits[1]}-${splits[0]}`;
  }

  /**
   * Returns a date string containing the current date
   * @returns Date string containing the current date in `dd.mm.yyyy` format
   */
  public static getCurrentDate(): DateString {
    const date: Date = new Date();
    const day: number = date.getUTCDate();
    const month: number = date.getUTCMonth() + 1;
    const year = String(date.getUTCFullYear());

    const dayPrefix: string = day < 10 ? '0' : '';
    const monthPrefix: string = month < 10 ? '0' : '';

    return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
  }
}
