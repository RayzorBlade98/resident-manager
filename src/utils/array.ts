/**
 * This module provides functionality regarding `Array` objects
 * like creating the cartesian product of multiple arrays.
 */

/**
 * Creates an array containing a sequence of numbers for a given range.
 *
 * Examples:
 * - `range(0, 4) := [0, 1, 2, 3, 4]`
 * - `range (0, 4, 2) := [0, 2, 4]`
 * - `range (0, 5, 2) := [0, 2, 4]`
 *
 * @param start start value of the sequence
 * @param stop end value of the sequence
 * @param step space between each element of the sequence (default: 1)
 * @returns sequence of numbers for the given range
 */
export function range(start: number, stop: number, step: number = 1): number[] {
  return [...Array(Math.floor((stop - start) / step) + 1)].map(
    (_, i) => start + i * step,
  );
}

/**
 * Creates the cartesian product for any number of given array.
 *
 * Examples:
 * - `cartesianProduct([1, 2], [3, 4]) := [[1, 3], [1, 4], [2, 3], [2, 4]]`
 * - `cartesianProduct([1, 2], [3, 4], [5, 6]) := [[1, 3, 5], [1, 3, 6], [1, 4, 5], [1, 4, 6], [2, 3, 5], [2, 3, 6], [2, 4, 5], [2, 4, 6]]`
 *
 * @param allEntries any number of array that should be combined
 * @returns array containing the cartesian product over all given arrays
 */
export function cartesianProduct<T>(...allEntries: T[][]): T[][] {
  return allEntries.reduce<T[][]>(
    (results, entries) =>
      results
        .map((result) => entries.map((entry) => result.concat([entry])))
        .reduce((subResults, result) => subResults.concat(result), []),
    [[]],
  );
}
