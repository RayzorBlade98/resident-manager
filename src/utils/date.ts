// Remove after adding another export:
/* eslint-disable import/prefer-default-export */

/**
 * Converts the given date to its string representaion (format: DD.MM.YYYY)
 * @param date date that should be converted
 * @returns string representation of the date
 */
export function dateToString(date: Date): string {
  const day: number = date.getDate();
  const month: number = date.getMonth() + 1;
  const year = String(date.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
