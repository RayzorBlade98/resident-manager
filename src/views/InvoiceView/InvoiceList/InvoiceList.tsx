import {
  Box,
  Divider,
  List,
  ListItemButton,
  ListItemText,
} from '@mui/material';
import React, { useState } from 'react';
import { useRecoilState, useRecoilValue } from 'recoil';
import { invoiceViewSelectedInvoiceSelector } from '../states/invoice_view_state';
import CreateInvoiceModal from './CreateInvoiceModal/CreateInvoiceModal';
import Invoice from '_/models/invoice/invoice';
import invoiceState from '_/states/invoice/invoice.state';

const styles = {
  box: {
    height: '100%',
  },
  list: {
    height: '100%',
    overflowY: 'auto',
    padding: 0,
    borderRight: 1,
    borderColor: 'divider',
  },
  listItemButton: {
    height: '11.03%',
  },
};

/**
 * Component that displays a list of all invoices
 */
function InvoiceList(): JSX.Element {
  const [showModal, setShowModal] = useState(false);

  const invoices = useRecoilValue(invoiceState);
  const [selectedInvoice, setSelectedInvoice] = useRecoilState(
    invoiceViewSelectedInvoiceSelector,
  );

  return (
    <>
      <CreateInvoiceModal showModal={showModal} onCloseModal={() => setShowModal(false)} />
      <Box sx={styles.box}>
        <List sx={styles.list}>
          <ListItemButton
            onClick={() => setShowModal(true)}
            sx={styles.listItemButton}
          >
            <ListItemText primary="Neue Abrechnung" />
          </ListItemButton>
          <Divider />
          {invoices.map((invoice: Invoice, i) => (
            <>
              <ListItemButton
                selected={invoice.id === selectedInvoice?.id}
                onClick={() => setSelectedInvoice(invoice)}
                key={invoice.id}
                sx={styles.listItemButton}
              >
                <ListItemText
                  primary={`${invoice.start.toString()} - ${invoice.end.toString()}`}
                />
              </ListItemButton>
              {i !== invoices.length - 1 && <Divider />}
            </>
          ))}
        </List>
      </Box>
    </>
  );
}

export default InvoiceList;
