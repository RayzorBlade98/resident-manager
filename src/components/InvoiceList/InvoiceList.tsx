/* eslint-disable react/jsx-props-no-spreading */
import { StyleAttribute } from 'glamor';
import React from 'react';
import GenericList from '../GenericComponents/GenericList/GenericList';
// eslint-disable-next-line max-len
import GenericListElement from '../GenericComponents/GenericList/GenericListElement';
import { Invoice } from '_/types/invoice';

interface InvoiceListProps {
  /**
   * List of included invoices
   */
  invoices: Invoice[];

  /**
   * Currently selected invoice
   */
  selectedInvoice?: Invoice;

  /**
   * Callback when selecting an invoice
   */
  onSelectInvoice: (invoice: Invoice) => void;

  /**
   * Additional stlye of the list container
   */
  style?: StyleAttribute;
}

/**
 * Component that displays a list of provided invoices
 */
function InvoiceList(props: InvoiceListProps): JSX.Element {
  return (
    <GenericList style={props.style} className="invoiceList">
      {props.invoices.map((invoice: Invoice) => (
        <GenericListElement
          onClick={() => {
            props.onSelectInvoice(invoice);
          }}
          selected={invoice === props.selectedInvoice}
          key={`$invoice-${invoice.start.month}-${invoice.start.year}`}
        >
          {`${invoice.start.month} ${invoice.start.year}`}
        </GenericListElement>
      ))}
    </GenericList>
  );
}

export default InvoiceList;
