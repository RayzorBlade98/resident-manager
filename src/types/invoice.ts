import { MonthYear } from './date';

/**
 * Object containing information about an invoice
 */
export interface Invoice {
  /**
   * First month of the invoice
   */
  start: MonthYear;

  /**
   * Last month of the invoice
   */
  end: MonthYear;

  /**
   * Invoice information of all residents
   */
  residentInformation: { [residentId: string]: ResidentInvoiceInformation };
}

/**
 * Object containing invoice information for a single resident
 */
export interface ResidentInvoiceInformation {
  /**
   * Id of the resident
   */
  residentId: string;
}
