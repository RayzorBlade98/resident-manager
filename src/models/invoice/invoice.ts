import { DeductionType } from '../incidentals/ongoing_incidentals';
import ResidentInvoiceInformation from './resident_invoice';
import MonthYear from '_/extensions/date/month_year.extension';
import { CurrencyInCents } from '_/utils/currency/currency.utils';

/**
 * Object containing information about an invoice
 */
export default interface Invoice {
  /**
   * Unique id of the invoice
   */
  id: string;

  /**
   * First month of the invoice
   */
  start: MonthYear;

  /**
   * Last month of the invoice
   */
  end: MonthYear;

  /**
   * Invoice information of all ongoing incidentals
   */
  ongoingIncidentalsInformation: {
    [incidentalsId: string]: OngoingIncidentalsInvoiceInformation;
  };

  /**
   * Invoice information of all residents
   */
  residentInformation: { [residentId: string]: ResidentInvoiceInformation };
}

/**
 * Object containing invoice information for a single ongoing incidentals
 */
interface OngoingIncidentalsInvoiceInformation {
  /**
   * Id of the ongoing incidentals
   */
  incidentalsId: string;

  /**
   * Name of the incidentals
   */
  name: string;

  /**
   * Total cost of the incidentals for the whole invoice
   */
  totalCost: CurrencyInCents;

  /**
   * Deduction type of the incidentals
   */
  deductionType: DeductionType;
}
