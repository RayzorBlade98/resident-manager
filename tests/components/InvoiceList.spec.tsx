/* eslint-disable max-len, @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call, func-names */
/* eslint-disable @typescript-eslint/no-unused-vars */

import { RenderResult, render, fireEvent } from '@testing-library/react';
import { css } from 'glamor';
import React from 'react';
import InvoiceList from '_/components/InvoiceList/InvoiceList';
import { Month } from '_/types/date';
import { Invoice } from '_/types/invoice';
import { range } from '_/utils/array';
import InvoiceBuilder from '_tests/__test_utils__/builders/invoice_builder';

describe('ResidentList', () => {
  let renderResult: RenderResult;
  let invoices: Invoice[];
  let selectedInvoice: Invoice;
  let onSelectInvoiceMock: jest.Mock;

  beforeEach(() => {
    InvoiceBuilder.setStart({
      month: Month.January,
      year: 2023,
    });
    invoices = range(0, 4).map((_) => new InvoiceBuilder().build());
    selectedInvoice = invoices[0];
    onSelectInvoiceMock = jest.fn();

    renderResult = render(
      <InvoiceList
        invoices={invoices}
        selectedInvoice={selectedInvoice}
        onSelectInvoice={onSelectInvoiceMock}
        style={css({ backgroundColor: 'red' })}
      />,
    );
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  test('should have the right amount of children', () => {
    // Assert
    const expectedChildren = 2 * invoices.length - 1;
    const actualChildren = renderResult.container.firstElementChild!.children.length;
    expect(actualChildren).toEqual(expectedChildren);
  });

  test('should call onSelectResident callback', () => {
    // Act
    const selectedIndex = 1;
    const listElement = renderResult.container.firstElementChild!.children.item(
      2 * selectedIndex,
    )!;
    fireEvent.click(listElement);

    // Assert
    expect(onSelectInvoiceMock).toHaveBeenCalledTimes(1);
    expect(onSelectInvoiceMock).toHaveBeenCalledWith(invoices[selectedIndex]);
  });

  test('should match snapshot', () => {
    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });
});
