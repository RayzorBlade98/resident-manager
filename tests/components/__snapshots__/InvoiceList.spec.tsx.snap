// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResidentList should match snapshot 1`] = `
.css-0,
[data-css-0] {
  width: 100%;
  height: 10%;
  color: #163172;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.css-0:hover,
[data-css-0]:hover {
  background-color: #c8daea;
  cursor: pointer;
}

.css-2,
[data-css-2] {
  width: 100%;
  height: 10%;
  color: #163172;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #c8daea;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -webkit-box-align: center;
  -webkit-align-items: center;
}

.css-2:hover,
[data-css-2]:hover {
  background-color: #c8daea;
  cursor: pointer;
}

.css-4,
[data-css-4] {
  background-color: red;
  overflow-y: auto;
}

<div>
  <div
    class="genericList invoiceList"
    data-css-4=""
  >
    <div
      class="listElement selectedListElement"
      data-css-2=""
    >
      Januar 2023
    </div>
    <hr
      class="hr"
    />
    <div
      class="listElement"
      data-css-0=""
    >
      Februar 2023
    </div>
    <hr
      class="hr"
    />
    <div
      class="listElement"
      data-css-0=""
    >
      MÃ¤rz 2023
    </div>
    <hr
      class="hr"
    />
    <div
      class="listElement"
      data-css-0=""
    >
      April 2023
    </div>
    <hr
      class="hr"
    />
    <div
      class="listElement"
      data-css-0=""
    >
      Mai 2023
    </div>
  </div>
</div>
`;
